### PR TITLE
Add support for PHP 8.1 & remove circular dependency with concrete implementation

### PR DESCRIPTION
**Pull request by arueckauer**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This PR

- adds PHP 8.1 support
- drops PHP 7.3 support
- removes circular dependencies with
  - laminas/laminas-auradi-config
  - laminas/laminas-pimple-config
  - laminas/laminas-servicemanager
- removes test classes
  - AuraDiTest (duplicated from laminas/laminas-auradi-config)
  - PimpleTest (duplicated from laminas/laminas-pimple-config)
  - ServiceManagerTest (duplicated from laminas/laminas-servicemanager)
- upgrades dependencies to new supported PHP version range
  - `laminas/laminas-coding-standard` to `2.3`
  - `laminas/laminas-servicemanager` to `3.10`
  - `phpunit/phpunit` to `9.5`

### CODE REVIEW

File: test/ServiceManagerTest.php [hunk 1]
```diff
@@ -14,9 +14,10 @@ class ServiceManagerTest extends AbstractMezzioContainerConfigTest
     use SharedTestTrait;
 
     /**
+     * @param array<string,mixed> $config
      * @return ServiceManager
      */
-    public function createContainer(array $config) : ContainerInterface
+    public function createContainer(array $config): ContainerInterface
     {
         return new ServiceManager($config);
```
Comment (Ocramius) [inline test/ServiceManagerTest.php]: I think the parameter `array<string, mixed> $config` is a bit too lax here, heh...

**Review by Ocramius — CHANGES_REQUESTED** [review]
Requires fixing CI failures on various pipeline matrix entries, sorry :|

File: test/ServiceManagerTest.php (hunk 1 above)
Comment (boesing) [reply test/ServiceManagerTest.php]: https://github.com/laminas/laminas-servicemanager/blob/3.11.x/src/ConfigInterface.php

We could the psalm type from laminas-servicemanager.
Would require a version bump here but that should be fine imho.

**Comment by boesing** [comment]
As already mentioned in slack:

Remove the pimple dependency here. This package should not have any dependency to a concrete implementation.
The container implementing component should require this package, not the other way around.

You don't want to get broken builds in here due to a bug in the pimple integration component.

BTW: I also failed in https://github.com/laminas/laminas-cache-storage-adapter-test - same issue at some point in time. but luckily, the `memory` adapter is almost the easiest one to upgrade to new PHP versions as there is no external dependency needed.

File: test/ServiceManagerTest.php (hunk 1 above)
Comment (arueckauer) [reply test/ServiceManagerTest.php]: I have tried setting the `@psalm-param` as defined in `ServiceManager`. But this is more specific than the method definition in `AbstractContainerTest`, which is also used by `AuraDiTest`. The configuration object of AuraDi is also not compatible with that of the config array for `ServiceManager`. Not really sure, how to solve this.

Comment (boesing) [reply test/ServiceManagerTest.php]: So `AuraDi` integration test is in this component as well? 

IMHO: kick all integration tests and instead add them to the corresponding components. No integration tests should be needed in this component.
We can, ofc, create a fake integration to have integration tests.
But yes, if we add psalm-types, ofc, anyone has to match these.
So when moving integration tests to their integrations, the test cases have to adapt as well.

**Comment by arueckauer** [comment]
All circular development dependencies with concrete implementations are removed. Delete tests are already present in concrete implementations. There are no tests left and thus CI is failing. How should this be handled? Remove PHPUnit or add rudimentary tests or ...?

**Comment by boesing** [comment]
We should not drop phpunit here.
There are files which are provided to the several implementations and thus we should at least verify that they test what we expect they should test.

Maybe we get some inspiration from here:
https://github.com/laminas/laminas-cache-storage-adapter-test/tree/2.0.x/test

However, this is just my opinion about these integration tests. Maybe @Ocramius or @weierophinney do have thoughts here.
I'd prefer to have some tests available but not sure if thats all required for a PHP 8.1 support PR...

File: composer.json [hunk 2]
```diff
@@ -23,15 +23,12 @@
...
         "psr/container": "^1.0"
     },
     "require-dev": {
-        "laminas/laminas-auradi-config": "^1.0.1",
-        "laminas/laminas-coding-standard": "~1.0.0",
-        "laminas/laminas-pimple-config": "^1.1",
-        "laminas/laminas-servicemanager": "^3.3.2",
-        "phpunit/phpunit": "^7.5.15 || ^8.3.4",
+        "laminas/laminas-coding-standard": "^2.3",
+        "phpunit/phpunit": "^9.5",
```
Comment (weierophinney) [inline composer.json]: Change this to `^9.5.5`. That's the first version to work with PHP 8.1, which _might_ help with the testing situation.

**Review by weierophinney — CHANGES_REQUESTED** [review]
Just remove the `phpunit.xml.dist` file. IIRC, the CI won't generate jobs for PHPUnit if there is no PHPUnit configuration.

File: composer.json (hunk 2 above)
Comment (boesing) [reply composer.json]: if we remove phpunit config, we can remove phpunit dependency as well 🤷🏼‍♂️

Comment (arueckauer) [reply composer.json]: Changed to `^9.5.5`.

**Comment by arueckauer** [comment]
@weierophinney You are right, removing `phpunit.xml.dist` does prevent CI pipeline to trigger unit tests. However, phpcs and psalm fail due to missing `test` directory. Instead of removing test setup completely and possibly being overseen to re-add coding style checks and static analysis, I implemented a test. This should solve the issue here.

**Comment by Ocramius** [comment]
@arueckauer psalm and phpcs use their respective configs: if the `test` directory does not exist, you can drop it from the config

**Comment by arueckauer** [comment]
@Ocramius I am aware of it and tried to explain in my last comment. Removing whole test setup is a possibility with the chance of not updating phpcs and psalm configuration as soon as tests are added again. I chose the other way and provided a test case. That way the "standard setup" is still in tact. Let me know if you want it changed. Otherwise this should be ready to be merged.

**Review by boesing — APPROVED** [review]
LGTM

**Pull request merged**